src/plugins/_manager.ts

```diff
@@ -0,0 +1,171 @@
+import * as child from 'child_process';
+import * as os from 'os';
+import * as path from 'path';
+import * as url from 'url';
+import * as fs from 'fs-extra';
+
+const MODULE_NOT_FOUND_ERROR_CODE = 'MODULE_NOT_FOUND'; // TODO is there a known constant we can use here?
+const MODULE_VERSION_MISMATCH_ERROR_CODE = 'MODULE_VERSION_MISMATCH';
+
+/**
+ * Error indicating a module was found but with an unexpected version.
+ */
+export class ModuleVersionMismatchError extends Error {
+
+  public readonly code: string = MODULE_VERSION_MISMATCH_ERROR_CODE;
+
+  constructor(pkg: string, expected: string, actual: string) {
+    super(`Version mismatch for package ${pkg}. Found ${actual} but requested ${expected}`);
+  }
+}
+
+export interface Plugin {
+
+  readonly instance: any;
+  readonly class: string;
+  readonly module: any;
+  readonly pkg: string;
+  readonly version: string;
+  readonly path: string;
+
+}
+
+export interface Package {
+
+  readonly module: any;
+  readonly pkg: string;
+  readonly version: string;
+  readonly path: string;
+
+}
+
+export interface PluginManagerLoadOptions {
+  readonly pkg: string;
+  readonly version: string;
+  readonly class: string;
+  readonly installEnv?: { [key: string]: any };
+  readonly properties?: { [key: string]: any };
+}
+
+export class PluginManager {
+
+  private readonly dir: string;
+
+  constructor(dir: string) {
+    this.dir = dir;
+  }
+
+  public load(options: PluginManagerLoadOptions): Plugin {
+
+    const pkg = this.loadPackage(options.pkg, options.version, options.installEnv ?? {});
+
+    const clazz = pkg.module[options.class];
+    if (!clazz) {
+      throw new Error(`Unable to locate class '${options.class}' in package '${options.pkg}@${options.version}'. Are you sure you exported it?`);
+    }
+
+    return { instance: new clazz(options.properties ?? {}), class: options.class, ...pkg };
+  }
+
+  private loadPackage(pkg: string, version: string, installEnv: { [key: string]: any }): Package {
+
+    if (isNaN(parseInt(version.charAt(0)))) {
```

Comment:
It's a bit of a smell to use `parseInt` without specifying a base, as the default base is platform-dependent and can be surprising.

```suggestion
    if (isNaN(parseInt(version.charAt(0), 10))) {
```

Also - what is `version` supposed to be like? I guess you mean for it to be the npm package version? In this case, I have to break unfortunate news to you... `npm` accepts the following version range notations:
- `^1.2.3` (you got this covered!)
- `1.2.x`
- `1.2`
- `1`
- Possibly some more...

Reply:
Well damn. Good catch. Added `semver` as a dependency and validated using `semver.Range(v).range`.